### DISCUSSION TITLE
How to disable lsp inline error messages?

### DISCUSSION THREAD

**Discussion by Jie-Fang** [ayamir/nvimdots#703] [General]
In some cases, I don't need lsp inline error diagnostic messages. Too many error messages make my codes messy. I can run LspStop to stop LSP, but i prefer to configure the setting to disable error hints and remain lsp functionalities.

How can I disable these error hints?

**Comment by charliie-dev** [comment]
I'm not entirely sure about your goal, but maybe the idea proposed in #666 would solve some of your issues?

**Comment by charliie-dev** [comment]
Do u want to disable 
1. certain error messages from certain lsp or a null-ls source?
2. all error messages from certain lsp or a null-ls source?
3. all error messages from all lsps and null-ls sources?

it needs different degrees of efforts to accomplish each goal.

**Reply by Jie-Fang** [reply]
Hi @CharlesChiuGit , I'm not sure what it is. I attach an image to describe the inline error messages in my case.
![Screenshot 2023-05-04 130109](https://user-images.githubusercontent.com/10357969/236116039-44399686-1cc7-4375-bb4a-935428ab64e4.png)
The inline error messages are pink words at the end of "wrong" code line.

**Reply by charliie-dev** [reply]
It seems like you're using `python` right?
It seems like `pyright` didn't resolve installed modules in your python venv properly.
If u use `pyright` lsp for `python`, what's your settings for `pyright`?

Can u show what's in your `:LspInfo`?

**Reply by Jie-Fang** [reply]
yes, I'm using python and pyright is as lsp backend. Where's the setting? I didn't customize it and used the default setting. Actually, pyright works well on most of the python packages, like numpy, builtin modules. But in my projects, there're some packages which are python wrapper of c++ source codes using pybind by ourselves. Some are python packages which have no source code, but only c++ dynamic library. So pyright may not work well on these packages. It will show the error messages even if the codes are right.

What I want is just to disable the error message.

Here is the LspInfo output:
![Screenshot 2023-05-04 134905](https://user-images.githubusercontent.com/10357969/236121745-1c1509f4-5b40-4d51-bbb2-8282d4b61376.png)

**Reply by charliie-dev** [reply]
> Some are python packages which have no source code, but only c++ dynamic library. So pyright may not work well on these packages. It will show the error messages even if the codes are right.

Seems reasonable, but I fount some interesting material that might helps.[^1][^2]
It would take more work, but it much better in the long run.

> What I want is just to disable the error message.

To disable `typecheck` and `importcheck` of that **certain python project**, create `pyrightconfig.json`in your project root and add following lines.
`pyrightconfig.json`:
```json
{
    "venvPath": "/home/username/.virtualenvs/",
    "venv": "myenv",
    "reportMissingTypeStubs": "none",
    "reportOptionalSubscript": false,
    "reportOptionalMemberAccess": false
}
```

Try this and let us know it this works.

[ref](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#type-check-diagnostics-settings) for `pyrightconfig`

---

If u want to ignore only certain line of diagnostics from `pyright`, u can add a comment behind that line, like so: [^3]
```python
import missing_module import name # pyright: ignore
import missing_module import name # pyright: ignore[reportMissingImports]
```

[^1]: https://github.com/microsoft/pyright/issues/1202
[^2]: https://stackoverflow.com/questions/73069721/dynamic-c-library-in-python-cant-find-dylib-file
[^3]: https://stackoverflow.com/questions/68822660/how-do-you-ignore-specific-pyright-type-checks-by-project-file-line

**Reply by Jie-Fang** [reply]
Thanks! @CharlesChiuGit
But it's still a little bit complicated for me. I found this hint is provided by lspconfig. After searching the related issues, I fixed it by adding `vim.diagnostic.config({virtual_text = false})` in the `core/options.lua`. See https://github.com/neovim/nvim-lspconfig/issues/662 . Now the virtual texts disappear but it still has a way to hint me where may exist problems through underlining the wrong codes. So I can press `<leader>ld` to get specific info.
BTW, I don't know where I should add this line of code to make the nvim configuration more precisely. Correct me if there's a better place to put this line of code.

**Reply by charliie-dev** [reply]
Wow, that's will disable all `virtual_text`, do u really sure u want that?
But if this method works for u, then it's fine.

I provided three methods, have u tried any one?

**Reply by Jie-Fang** [reply]
It seems these pyright configurations are applied for all modules. I can't find a way to take effect on a specific module.
So these settings is the same as what I have done previously. But my method doesn't disable any pyright functionalities and I can still obtain the diagnostic info by looking at if there're any underlined codes or toggle lsp document diagnostics(`<leader>td`) or toggle lsp document diagnostics(`<leader>tw`). I just remove the virtual texts which makes my code messy.

**Reply by charliie-dev** [reply]
> It seems these pyright configurations are applied for all modules. I can't find a way to take effect on a specific module.

it's not possible.

**Reply by charliie-dev** [reply]
feel free to close this discussion if that works for u.

**Comment by ayamir** [comment]
@Jie-Fang Set it in `core/settings.lua` now. FYI: 8e465550110bec414e926e489a58c4665ddc1931
